File: LeetCode/20. Valid Parentheses.js
// https://leetcode.com/problems/valid-parentheses/description/
const { test } = require('../test');

const map = { '(': ')', '{': '}', '[': ']' };
const opens = new Set(Object.keys(map));

const isValid = (str) => {
  const stack = [];

  for (let i = 0; i < str.length; i++) {
    const current = str[i];

    if (opens.has(current)) {
      stack.push(current);

      continue;
    }

    const top = stack.pop();

    if (map[top] !== current) {
      return false;
    }
  }

  return stack.length === 0;
};

test(isValid, [
  {
    input: ['()'],
    expected: true,
  },
  {
    input: ['()[]{}'],
    expected: true,
  },
  {
    input: ['(]'],
    expected: false,
  },
  {
    input: ['([])'],
    expected: true,
  },
  {
    input: ['([)]'],
    expected: false,
  },
]);

File: LeetCode/20. Valid Parentheses.py
# https://leetcode.com/problems/valid-parentheses/description/
from functools import cached_property
from test import test
from typing import List


class Solution:
    my_dict = {
        "(": ")",
        "{": "}",
        "[": "]",
    }

    @cached_property
    def opens(self):
        return set(self.my_dict.keys())

    def isValid(self, s: str) -> bool:
        stack: List[str] = []

        for char in s:
            if char in self.opens:
                stack.append(char)
                continue

            if len(stack) == 0:
                return False

            last = stack.pop()

            if self.my_dict[last] != char:
                return False

        return len(stack) == 0


solution = Solution()


test(
    solution.isValid,
    [
        {
            "input": ["()"],
            "expected": True,
        },
        {
            "input": ["()[]{}"],
            "expected": True,
        },
        {
            "input": ["(]"],
            "expected": False,
        },
        {
            "input": ["([])"],
            "expected": True,
        },
        {
            "input": ["([)]"],
            "expected": False,
        },
    ],
)
